---
title: "autonomics"
subtitle: "Generic & intuitive cross-platform data analysis"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{using_autonomics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=TRUE, echo=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse = TRUE, cache = TRUE)
#str(knitr::opts_chunk$get())
```
<!-- ![ ](https://bitbucket.org/graumannlabtools/autonomics/downloads/read_prepro_analyze.png) -->

# Introduction

`autonomics` is created to make omics analysis *generic*, *intuitive*, and *easy to use/maintain*, providing oneliner importers for *RNAseq*, *MS proteomics*, *SOMAscan* and *Metabolon* platforms, as well as a generic importer for data from *any rectangular omics file*. 

These oneliners (1) import data, (2) merge sample/feature information, (3) perform platform-aware preprocessing, (4) impute consistent nondetects, (5) infer the design from systematically provided sample names, (6) pca/lda/sma/pls explore, (7) fit a linear model, (8) extract contrasts of interest, and (9) visualize the results.

The customizable `formula/block/contrast` interface allows for use on *any dataset*, fitting *any general linear model*, extracting *any contrast/coefficient*, using *random effects* or *precision weights* if required, building on top of the powerful `limma` platform for statistical analysis.

This package also provides intuitive functions to analyze *complex designs*. Particularly interesting is the `plot_contrastogram` (contrast + diagram) function, which creates a powerful summary of the differential expression structure, and is particulary useful in complex datasets.

With repetitive parts automated, common steps generified, and common operations functionalized, cross-platform omics data analysis can become an intuitive experience. 

Enjoy!

```{r}
# Load 
require(autonomics)
```

# Read / Analyze

## RNAseq {-}

The common workflow in RNAseq data analysis is to
(1) take reads from fastq files
(2) align them to a reference genome, producing BAM files
(3) count the number of reads per gene, producing a count file.
(4) normalize the counts per million reads (cpm), accounting for library size/composition
(5) fit a linear model, using either a binomial distribution (edgeR/DEseq) or precision weights to account for the properties of count data.

Alignment (hours) and (to a lesser extent) counting are time consuming processes, performed generally only once, prior to data analysis. 
RNAseq data analysis then typically starts from BAM or count files, for both of which autonomics provides dedicated importers:

```{r}
# BILLING 2019
     # Read
         file <- download_data('billing19.rnacounts.txt')
         object <- .read_rnaseq_counts(file)
     # Read / Analyze
         object <-  read_rnaseq_counts(file, voom=FALSE, plot=FALSE)
     # Read / Weight / Analyze
         object <-  read_rnaseq_counts(file)
         
# GSE161731
    # Download
        require(magrittr)
        basedir <- '~/autonomicscache/datasets'
        subdir  <- '~/autonomicscache/datasets/GSE161731'
        if (!dir.exists(subdir))  GEOquery::getGEOSuppFiles(
                                        "GSE161731",baseDir=basedir)
        file  <- paste0(subdir,'/GSE161731_counts.csv.gz')
        sfile <- paste0(subdir,'/GSE161731_counts_key.csv.gz')
    # Read
        object <- .read_rnaseq_counts(file, sfile = sfile, sfileby ='rna_id',
                                       subgroupvar = 'gender')
    # Read / Analyze
        object <- read_rnaseq_counts(file, sfile = sfile, sfileby = 'rna_id',
                                       subgroupvar = 'gender', voom = FALSE)
    # Read / Weight / Analyze
        object <- read_rnaseq_counts(file, sfile = sfile, sfileby ='rna_id',
                                       subgroupvar = 'gender', plot = TRUE)
    # Read / Weight / Block / Analyze (dupcor takes some time)
        #object <- read_rnaseq_counts(file, sfile = sfile, sfileby='rna_id',
        #                             subgroupvar='gender', block='subject_id')

#file <- download_data('billing16.rnacounts.txt')
#object <- read_rnaseq_counts(file)
```

## RNAseq BAM {-}

```{r}
# BILLING 2016
    # Read
        file <- download_data('billing16.bam.zip')
        object <- .read_rnaseq_bams(file, paired = TRUE, genome = 'hg38', plot = FALSE)
    # Read / Analyze
        object <- read_rnaseq_bams(file, paired = TRUE, genome = 'hg38', voom = FALSE, plot = FALSE)
    # Read / Analyze / weight
        object <- read_rnaseq_bams(file, paired = TRUE, genome = 'hg38')
```


```{r, fig.width=7, fig.height=4, out.width="95%"}
require(autonomics, quietly = TRUE)
file <- download_data('halama18.metabolon.xlsx')
object <- read_metabolon(file) 
```

## Metabolon {-}
                             
```{r}
file <- download_data('halama18.metabolon.xlsx')
object <- read_metabolon(file)
```


## Somascan {-}

```{r}
file <- download_data('atkin18.somascan.adat')                  # SOMAscan adat
object <- read_somascan(file)
```

## Proteingroups {-}

```{r}
proteingroups <- download_data('billing19.proteingroups.txt')
select_subgroups <- c('E00_STD', 'E01_STD', 'E02_STD',
                        'E05_STD', 'E15_STD', 'E30_STD', 'M00_STD')
object <- read_proteingroups(proteingroups, select_subgroups = select_subgroups)
```

## Phosphosites {-}

```{r}
phosphosites <- download_data('billing19.phosphosites.txt')
object <- read_phosphosites(phosphosites, proteingroups, select_subgroups = select_subgroups)
```


# Transform

# Explore

# Contrast 

# Connect



# References

A. Halama, M. Kulinski, S. Dib, S.B. Zaghlool, K.S. Siveen, A. Iskandarani, J. Zierer 3, K.S. Prabhu, N.J. Satheesh, A.M. Bhagwat, S. Uddin, G. KastenmÃ¼ller, O. Elemento, S.S. Gross, K. Suhre (2018). Accelerated lipid catabolism and autophagy are cancer survival mechanisms under inhibited glutaminolysis. Cancer Lett., 430:133-147. doi:10.1016/j.canlet.2018.05.017









