---
title: "autonomics"
subtitle: "generic and intuitive cross-platform data analysis"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{using_autonomics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=TRUE, echo=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse = TRUE, cache = TRUE)
#str(knitr::opts_chunk$get())
```
<!-- ![ ](https://bitbucket.org/graumannlabtools/autonomics/downloads/read_prepro_analyze.png) -->




```{r, eval=FALSE}
# Install
    install.packages('remotes')
    remotes::install_github('bhagwataditya/autonomics',  
                             repos = BiocManager::repositories(), 
                             dependencies = TRUE, upgrade = FALSE)
                             
# Run
    file <- download_data('billing16.bam.zip')                      # RNAseq BAM
    read_bam(file, paired = TRUE, genome = 'hg38')
    
    file <- download_data('billing16.rnacounts.txt')                # RNAseq counts
    read_counts(file)
        
    file <- download_data('halama18.metabolon.xlsx')                # Metabolon xlsx
    read_metabolon(file)
    
    rm_subgroups <- c('BLANK_BM00', 'BM00_BM00', 'EM01_EM00', 'EM05_EM02', 'EM30_EM15')
    proteingroups <- download_data('billing19.proteingroups.txt')   # LCMSMS Proteingroups
    read_proteingroups(proteingroups, rm_subgroups = rm_subgroups)
    
    phosphosites <- download_data('billing19.phosphosites.txt')     # LCMSMS Phosphosites
    read_phosphosites(phosphosites, proteingroups, rm_subgroup = rm_subgroups)
    
    file <- download_data('atkin18.somascan.adat')                  # SOMAscan adat
    read_somascan(file)
```

# Background: promises and challenges

Technological breakthroughs have made high-throughput cross-platform omics profiling a reality, opening fascinating possibilities to unravel the depths of biology and personalized medicine. Successful cross-platform omics data analysis, however, remains a challenging endeavour. Firstly,  platforms and the properties of their data differ, and proper platform-aware preprocessing is required if analysis is to be successful. Secondly, 


## Platform-aware preprocessing

Firstly, the different platforms use very different technologies, and this leads to data with different properties. Successful cross-platform omics data asnalysis requires for data from each of thes platforms to be preprocessed fittingly prior to statistical analysis.

**RNAseq** transcript counts are heteroscedastic (higher variances for lower counts), requiring proper statistical adjustment to allow for meaningful analysis. This statistical adjustment is better when performed in a design-aware fashion, ideally incorporating information on both fixed as well as random effects. Furthermore, RNAseq counts need to be normalized for library size/composition for powerful analysis.

**MS proteomics** platforms are plagued by a large number of *nondetects*, representing a mix of true non-detection (missing not at random) as well as stochastic dropout (missing at random), requiring proper nondetect handling for powerful analysis. MS proteomics also has the additional challenge of *protein group non-resolution*: groups of highly similar protein sequences cannot be resolved in identififcation. In some cases, the protein sequence similarity has a biological origin (isoforms and paralogs), and non-resolution is unavoidable (and even meaningful). In many other cases, the proteingroup non-resolution arises due to the presence of redundant, low-quality sequences in the protein sequence database used for protein identification, such as the popular species-specific isoform databases from UniProt. These databases are comprehensive and up-to-date, but they also contain many redundancies due to uncurated, low quality sequences, such as different fragments of the same protein sequence as separate entries. These low-quality sequences are neither meaningful nor useful, and resolving each proteingroup to a single relevant annotation is recommended to enable powerful analysis, visualization or pathway enrichment. *Alternative multiplexing workflows lead to alternative abundance representations*. Label-free quantification involves quantification without multiplexing: each sample is injected as a seperate run into the mass spectrometer, and protein intensities are computed. In multiplexed workflows mass labels are added to make subsequent samples slightly heavier, allowing them to be mixed/injected together, and protein abundances are expressed as sample ratios (rather than intensities), allowing the elimination of run effects. Intensities and ratios have different properties and need to be dealt with appropriately.

**MS phosphosite** profiling is challenged by the mixup of phosphorylation and protein expression, requiring deconvolution of the two processes to get the true phosphorylation signal.

**SOMAscan** plasma proteomics results are nearly ready-to-use, only requiring sample/feature filtering on type/quality prior to statistical analysis.

## Platform-aware import

Secondly, different data preprocessing pipelines produce data in different layouts, though, off course, composed of the same three pieces of information (abundances, sample ids/subgroups/covariates, and feature ids/names/annotations). 
**Metabolon** returns metabolite data in a multi-sheet xlsx document. Relevant data is packed in the second sheet, organized in a feature x sample (row x col) fashion, with sample ids/subgroups in CLIENT_IDENTIFIER/Group columns, and feature ids/names in COMPOUND_ID/BIOCHEMICAL column.
**SOMAscan** provides plasma proteomics data as a txt with a transposed layout: samples are in rows (rather than columns) and features in columns. Sample ids/subgroup are in SampleId/SampleGroup rows, and feature ids/names are in SeqId/Target columns.
**MaxQuant proteingroups/phosphosites** provide data in txt, with features in rows, sample data missing, and blocks of different measures spread out in the (about 100 or more) columns. Only a few of these columns are relevant for downstream analysis, and selecting the right (e.g. Normalized Ratios rather than Intensities for SILAC-multiplexed data) is essential for successful subsequent analysis.
**RNAseq** preprocessing pipelines subsequently provide the data as fastq (read sequences), bam (aligned read sequences) or count files (number of read sequences per gene).


## Study-aware statistical analysis

Thirdly, different studies have different designs, requiring different statistical modeling and contrast analysis. 
Some studies are simple and involve comparing the effect of some treatment against a control. Others involve multiple factors (e.g. time and concentration). Some contain repeated measures (e.g. of a subject). Others have batch effects (run day, hospital) or covariates (age, gender, bmi, ethnicity) that need to be accounted for. It is important to account properly for relevant batch effects or covariates, but it takes some skill to identify which variables to include into the model. The statistical model can be formulated in alternative ways, each of which lead to coefficients with different interpretations. For more complex designs formulating the contrasts to be examined itself can require some skill. 


## Generic software design

Fourthly, cross-platform integration, method innovation, and cognitive efficiency all ask for tidy software with a generic design. Omics integration requires a generic representation. Method innovation requires tidy software, easy to use but also easy to maintain and extend


## Balancing automation and interaction

Many aspects of data import, preprocessing, and statistical analysis can be automated, freeing cognitive resources which can then be used for creative biological discovery. At the same time, for creative exploration requires a balance of automation and interaction, allowing the former to generate ideas, which can then be creatively explored through latter.


# Results: autonomics

The autonomics package was created to fulfill these aims. Through highly automated importers which perform data import, preprocessing, 

It contains oneliner importers for data from RNAseq, MS Proteomics/Phosphoproteomics, SOMAscan and Metabolon platforms, as well as a generic importer which can be used for data from new platforms. Each of these importers is highly automated, performing platform-specific data import/extraction, preprocessing, as well as statistical analysis. Subgroup definitions are either extracted from data files (from dedicated columns in e.g. SOMAscan/Metabolon data), guessed from sample names (KD.R1 -> KD), or merged in from another file. In the following sections we discuss each of these importers in more detail.

## Generic: read_omics

## Metabolon: read_metabolon

**Metabolon** data are imported from the relevant excel worksheet with little additional preprocessing.

## SOMAscan: read_somascan

**SOMAscan** plasma proteomics data are imported and features/samples are filtered on quality and type. Sample type filtering includes removing hybridization/elution/non-human controls. Feature type filtering includes removal of Calibrators, QC features, and Buffer features. Feature and sample filtering includes removing "FAIL"-ed features/samples. 

## RNAseq: read_bam/read_counts

**RNAseq counts** are TMM-normalized (correcting for library size/composition) and voom precision weights are computed to correct for the mean/variance relationship of count data. 

## Proteomics/Phosphoproteomics: read_proteingroups/read_phosphosites

Automated importers are MaxQuant-preprocessed **MS proteomics** data in proteingroups/phosphosites files are imported by extracting the relevant columns (out of about 100), imputing groupwise nondetects (NA in all replicates of a subgroup) while leaving other nondetects NA, resolving proteingroups by dropping lower-quality sequences, guessing the multiplexing scheme from column names, and simplifying names of abundance columns. For **phosphosite** data occupancies are additionally computed, allowing a separation of phosphorylation and protein expression. 





It lifts the burden of platform/study specifics by automating data import, preprocessing, and statistical analysis. Its tidy, functional programming design at the same time allows for easy interaction, maintenance, and extension.
In this section, autonomics is first demonstrated on Metabolon profiling data from a complex design.
Next, examples in proteomics/phosphoproteomics as well as RNAseq are given to show the specifics of each of these platforms.  Finally, it is shown how autonomics can be used for any dataset from any omics platform.



lift this burden. It aims to make cross-platform omics data analysis generic and intuitive. Platform specifics are eliminated by e.g. mean/variance modeling RNAseq counts, imputing groupwise NAs in MS proteomics data, and computing occupancies for phosphosite intensities. Statistical modeling is simplified by automatically creating design matrix from data file or sample names. 



SO MANY PAPERS SO MANY METHODS WHICH TO USE (E.G. NORMALIZATION)

NORMALIZATION NEEDS TO BE DESIGN-AWARE

UNIFIED INTERFACE: SUBGROUP + COVARIATES + BLOCK




(e.g. accounting for the mean variance relationship in RNAseq counts, imputing systematic NAs in proteomics data), dataset specifics (by extracting design info from data file or sample names and formulating a fitting statistical model)


(eliminating platform/design specifics) and intuitive (unified interface/automated visualization), and therefore, enjoyable. In the following sections, we will discuss for each of the major platforms how autonomics can be used 




By automating repetitive tasks, cognitive resources can 



# Read

autonomics provides dedicated importers for data from:

* Transcriptomics
    - RNAseq:      `read_bam`, `read_counts` 
    - Microarrays: `read_affymetrix`
* Proteomics: 
    - SOMAscan `read_somascan`
    - Mass Spectrometry: `read_proteingroups`, `read_phosphosites`, after MaxQuant.
* Metabolomics: 
    - Metabolon: `read_metabolon`
  
These importers use the generic backend `read_omics`, which can be used to import rectangular data from any new platform, by specifying the location of the three rectangular components in the data file: `fdata` (feature data), `sdata` (sample data), and `exprs` (expression values), which are extracted and wrapped into a `SummarizedExperiment` with feature_id and sample_id values in fdata and sdata.


![](https://bitbucket.org/graumannlabtools/autonomics/downloads/importers.png)

# Preprocess

Each importer performs all preprocessing steps required to get the data ready for statistical analysis, as shown in the table below.  

* **Sample** names are *demultiplexed* (proteingroups), technical/failing samples are *removed*, the *subgroup* is extracted, and further decomposed into *subfactors*.  
* **Features** are *aggregated*, technical/unexpressed features are *removed*, and feature identifications are *simplified* (proteomics).
* **Exprs** are transformed: (labeled proteingroups) ratios can be inverted, samples normalized, epxrs log2 transformed, precision weights computed (RNAseq), consistent non-detects imputed (proteingroups), and occupancies computed (phosphosites).


![](https://bitbucket.org/graumannlabtools/autonomics/downloads/prepro.png)

# Explore

After these steps, the data is ready for statistical analysis. As a first analysis step, each of the importers plots the  principal component analysis biplot. The subgroup extraction and decomposition in the previous step is used for coloring: the first subfactor is mapped to color hue and the second to luminance. This allows to get an easy-to-comprehend overview of the dataset, as shown below for the glutaminase dataset from Halama et al. (2018), which investigates the effect of different concentrations of glutaminase inhibitor (Untreated, Vehicle, 5 uM, 10 uM) at different time points (10h, 24h, 48h, 72h)


```{r, fig.width=7, fig.height=4, out.width="95%"}
require(autonomics, quietly = TRUE)
file <- download_data('halama18.metabolon.xlsx')
object <- read_metabolon(file) 
```

# Transform

# Explore

# Contrast 

# Connect



# References

A. Halama, M. Kulinski, S. Dib, S.B. Zaghlool, K.S. Siveen, A. Iskandarani, J. Zierer 3, K.S. Prabhu, N.J. Satheesh, A.M. Bhagwat, S. Uddin, G. Kastenmüller, O. Elemento, S.S. Gross, K. Suhre (2018). Accelerated lipid catabolism and autophagy are cancer survival mechanisms under inhibited glutaminolysis. Cancer Lett., 430:133-147. doi:10.1016/j.canlet.2018.05.017









