<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>importomics • importomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="importomics">
<meta property="og:description" content="importomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">importomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/using_importomics.html">importomics</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using_importomics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>importomics</h1>
            <h3 class="subtitle">Generifying and intuifying cross-platform omics analysis</h3>
            
      
      
      <div class="hidden name"><code>using_importomics.Rmd</code></div>

    </div>

    
    
<p><img src="https://bitbucket.org/graumannlabtools/importomics/downloads/read_prepro_analyze.png"></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Install</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">'remotes'</span><span class="op">)</span>
    <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">'bhagwataditya/importomics'</span>,  
                             repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html">repositories</a></span><span class="op">(</span><span class="op">)</span>, 
                             dependencies <span class="op">=</span> <span class="cn">TRUE</span>, upgrade <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
                             
<span class="co"># Run</span>
    <span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing16.bam.zip'</span><span class="op">)</span>                      <span class="co"># RNAseq BAM</span>
    <span class="fu"><a href="../reference/read_bam.html">read_bam</a></span><span class="op">(</span><span class="va">file</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>, genome <span class="op">=</span> <span class="st">'hg38'</span><span class="op">)</span>
    
    <span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing16.rnacounts.txt'</span><span class="op">)</span>                <span class="co"># RNAseq counts</span>
    <span class="fu"><a href="../reference/read_counts.html">read_counts</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span>
        
    <span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'halama18.metabolon.xlsx'</span><span class="op">)</span>                <span class="co"># Metabolon xlsx</span>
    <span class="fu"><a href="../reference/.read_metabolon.html">read_metabolon</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span>
    
    <span class="va">rm_subgroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'BLANK_BM00'</span>, <span class="st">'BM00_BM00'</span>, <span class="st">'EM01_EM00'</span>, <span class="st">'EM05_EM02'</span>, <span class="st">'EM30_EM15'</span><span class="op">)</span>
    <span class="va">proteingroups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing19.proteingroups.txt'</span><span class="op">)</span>   <span class="co"># LCMSMS Proteingroups</span>
    <span class="fu"><a href="../reference/read_proteingroups.html">read_proteingroups</a></span><span class="op">(</span><span class="va">proteingroups</span>, rm_subgroups <span class="op">=</span> <span class="va">rm_subgroups</span><span class="op">)</span>
    
    <span class="va">phosphosites</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing19.phosphosites.txt'</span><span class="op">)</span>     <span class="co"># LCMSMS Phosphosites</span>
    <span class="fu"><a href="../reference/read_proteingroups.html">read_phosphosites</a></span><span class="op">(</span><span class="va">phosphosites</span>, <span class="va">proteingroups</span>, rm_subgroup <span class="op">=</span> <span class="va">rm_subgroups</span><span class="op">)</span>
    
    <span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'atkin18.somascan.adat'</span><span class="op">)</span>                  <span class="co"># SOMAscan adat</span>
    <span class="fu"><a href="../reference/.read_somascan.html">read_somascan</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span></code></pre></div>
<div id="read" class="section level1">
<h1 class="hasAnchor">
<a href="#read" class="anchor"></a>Read</h1>
<p>Importomics provides dedicated importers for data from:</p>
<ul>
<li>Transcriptomics
<ul>
<li>RNAseq: <code>read_bam</code>, <code>read_counts</code>
</li>
<li>Microarrays: <code>read_affymetrix</code>
</li>
</ul>
</li>
<li>Proteomics:
<ul>
<li>SOMAscan <code>read_somascan</code>
</li>
<li>Mass Spectrometry: <code>read_proteingroups</code>, <code>read_phosphosites</code>, after MaxQuant.</li>
</ul>
</li>
<li>Metabolomics:
<ul>
<li>Metabolon: <code>read_metabolon</code>
</li>
</ul>
</li>
</ul>
<p>These importers use the generic backend <code>read_omics</code>, which can be used to import rectangular data from any new platform, by specifying the location of the three rectangular components in the data file: <code>fdata</code> (feature data), <code>sdata</code> (sample data), and <code>exprs</code> (expression values), which are extracted and wrapped into a <code>SummarizedExperiment</code> with feature_id and sample_id values in fdata and sdata.</p>
<p><img src="https://bitbucket.org/graumannlabtools/importomics/downloads/importers.png"></p>
</div>
<div id="preprocess" class="section level1">
<h1 class="hasAnchor">
<a href="#preprocess" class="anchor"></a>Preprocess</h1>
<p>Each importer performs all preprocessing steps required to get the data ready for statistical analysis, as shown in the table below.</p>
<ul>
<li>
<strong>Sample</strong> names are <em>demultiplexed</em> (proteingroups), technical/failing samples are <em>removed</em>, the <em>subgroup</em> is extracted, and further decomposed into <em>subfactors</em>.<br>
</li>
<li>
<strong>Features</strong> are <em>aggregated</em>, technical/unexpressed features are <em>removed</em>, and feature identifications are <em>simplified</em> (proteomics).</li>
<li>
<strong>Exprs</strong> are transformed: (labeled proteingroups) ratios can be inverted, samples normalized, epxrs log2 transformed, precision weights computed (RNAseq), consistent non-detects imputed (proteingroups), and occupancies computed (phosphosites).</li>
</ul>
<p><img src="https://bitbucket.org/graumannlabtools/importomics/downloads/prepro.png"></p>
</div>
<div id="explore" class="section level1">
<h1 class="hasAnchor">
<a href="#explore" class="anchor"></a>Explore</h1>
<p>After these steps, the data is ready for statistical analysis. As a first analysis step, each of the importers plots the principal component analysis biplot. The subgroup extraction and decomposition in the previous step is used for coloring: the first subfactor is mapped to color hue and the second to luminance. This allows to get an easy-to-comprehend overview of the dataset, as shown below for the glutaminase dataset from Halama et al. (2018), which investigates the effect of different concentrations of glutaminase inhibitor (Untreated, Vehicle, 5 uM, 10 uM) at different time points (10h, 24h, 48h, 72h)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://bitbucket.org/graumannlabtools/autonomics">importomics</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'halama18.metabolon.xlsx'</span><span class="op">)</span>
<span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_metabolon.html">read_metabolon</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span> 
<span class="co">##  Read: ~/importomicscache/datasets/halama18/halama18.metabolon.xlsx</span>
<span class="co">##      Infer design from sample_ids</span>
<span class="co">##      Guess replicate values from sampleids: ctrl_h10_R1 =&gt; R1</span>
<span class="co">##      Log2 transform</span>
<span class="co">##  Add PCA</span>
<span class="co">##      Retain 391/391 features: non-zero, non-NA, and non-NaN for some sample</span>
<span class="co">##      Run limma</span>
<span class="co">##      Retain 96/96 samples: ~!is.na(~subgroup)</span></code></pre></div>
<p><img src="using_importomics_files/figure-html/unnamed-chunk-3-1.png" width="95%"></p>
</div>
<div id="transform" class="section level1">
<h1 class="hasAnchor">
<a href="#transform" class="anchor"></a>Transform</h1>
</div>
<div id="explore-1" class="section level1">
<h1 class="hasAnchor">
<a href="#explore-1" class="anchor"></a>Explore</h1>
</div>
<div id="contrast" class="section level1">
<h1 class="hasAnchor">
<a href="#contrast" class="anchor"></a>Contrast</h1>
</div>
<div id="connect" class="section level1">
<h1 class="hasAnchor">
<a href="#connect" class="anchor"></a>Connect</h1>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>A. Halama, M. Kulinski, S. Dib, S.B. Zaghlool, K.S. Siveen, A. Iskandarani, J. Zierer 3, K.S. Prabhu, N.J. Satheesh, A.M. Bhagwat, S. Uddin, G. Kastenmüller, O. Elemento, S.S. Gross, K. Suhre (2018). Accelerated lipid catabolism and autophagy are cancer survival mechanisms under inhibited glutaminolysis. Cancer Lett., 430:133-147. <a href="doi:10.1016/j.canlet.2018.05.017" class="uri">doi:10.1016/j.canlet.2018.05.017</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aditya Bhagwat, Johannes Graumann, Shahina Hayat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
