<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>autonomics • autonomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="autonomics">
<meta property="og:description" content="autonomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">autonomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/using_autonomics.html">autonomics</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bhagwataditya/autonomics/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using_autonomics_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>autonomics</h1>
            <h3 class="subtitle">Read + preprocess + analyze + visualize cross-platform omics data</h3>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bhagwataditya/autonomics/blob/master/vignettes/using_autonomics.Rmd"><code>vignettes/using_autonomics.Rmd</code></a></small>
      <div class="hidden name"><code>using_autonomics.Rmd</code></div>

    </div>

    
    
<!-- ![ ](https://bitbucket.org/graumannlabtools/autonomics/downloads/read_prepro_analyze.png) -->
<div id="abstract" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p><code>autonomics</code> is created to make cross-platform omics analysis intuitive and enjoyable. It contains <strong>import + preprocess + analyze + visualize</strong> one-liners for <em>RNAseq</em>, <em>MS proteomics</em>, <em>SOMAscan</em> and <em>Metabolon</em> platforms and a <em>generic</em> importer for data from <em>any rectangular omics file</em>. With a focus on not only automation but also customization, these importers have a flexible <code>formula</code>/ <code>block</code>/<code>contrastdefs</code> interface which allows to define any statistical formula, fit any general linear model, quantify any contrast, and use random effects or precision weights if required, building on top of the powerful <code>limma</code> platform for statistical analysis. It also offers exquisite support for analyzing <strong>complex designs</strong> such as the innovative <strong>contrastogram</strong> visualization to unravel and summarize the differential expression structure in complex designs. By autonomating repetitive tasks, generifying common steps, and intuifying complex designs, it makes cross-platform omics data analysis a fun experience. Try it and enjoy :-).</p>
</div>
<div id="read-preprocess-analyze-visualize" class="section level1">
<h1 class="hasAnchor">
<a href="#read-preprocess-analyze-visualize" class="anchor"></a>read + preprocess + analyze + visualize</h1>
<p>Autonomics offers <strong>import + preprocess + analyze + visualize</strong> one-liners for <em>RNAseq</em>, <em>MS proteomics</em>, <em>SOMAscan</em> and <em>Metabolon</em> platforms, as well a <em>generic</em> importer for data from <em>any rectangular omics file</em>. We will discuss each of these in more detail in the following sections, but all of them follow the following general steps:</p>
<ul>
<li>
<strong>read</strong>
<ul>
<li>exprs, features, samples</li>
<li>extract <strong>subgroups</strong> from sampleids / samplefile</li>
<li>wrap these into a <strong>SummarizedExperiment</strong>
</li>
</ul>
</li>
<li>
<strong>preprocess</strong>
<ul>
<li><strong>normalize</strong></li>
<li>
<strong>log2</strong> transform</li>
<li>
<strong>impute</strong> in a platform-aware fashion</li>
<li><strong>filter</strong></li>
</ul>
</li>
<li>
<strong>analyze</strong>
<ul>
<li>plot <strong>sample distributions</strong>
</li>
<li>plot <strong>detections</strong> (absent/present)</li>
<li>plot <strong>pca</strong> biplots</li>
<li>
<strong>fit</strong> the linear model <code>~0+subgroup</code>
</li>
<li>quantify <strong>contrasts</strong> between subsequent factor levels and <strong>plot volcanoes</strong>
</li>
</ul>
</li>
</ul>
<div id="rnaseq" class="section level2">
<h2 class="hasAnchor">
<a href="#rnaseq" class="anchor"></a>RNAseq</h2>
<p>Autonomics provides ready-to-use importers for both <strong>count</strong> as well as <strong>BAM</strong> files, which read / preprocess / analyze RNAseq data. Specific to RNAseq is counting reads using Rsubread::featureCounts (for read_rna_bams)), normalizing for library composition (<strong>cpm</strong>) or gene length (<strong>tpm</strong>), estimating <strong>voom</strong> precision weights, and using these in linear modeling. Let’s illustrate both of these on example datasets:</p>
<div id="read_rnaseq_counts" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#read_rnaseq_counts" class="anchor"></a>read_rnaseq_counts</h3>
<p><strong>Billing et al. (2019)</strong> studied the differentiation of embryonic (<strong>E00</strong>) into mesenchymal stem cells (<strong>E00.8</strong>, <strong>E01</strong>, <strong>E02</strong>, <strong>E05</strong>, <strong>E15</strong>, <strong>E30</strong>), with similar properties as bone-marrow mesenchymal stem cells (<strong>M00</strong>). Importing the RNAseq counts:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bhagwataditya/autonomics">autonomics</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">## Attache Paket: 'autonomics'</span>
<span class="co">## The following object is masked from 'package:stats':</span>
<span class="co">## </span>
<span class="co">##     biplot</span>
<span class="va">object</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/.read_rnaseq_bams.html">read_rnaseq_counts</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing19.rnacounts.txt'</span><span class="op">)</span>, 
                              pca <span class="op">=</span> <span class="cn">TRUE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">##  Read C:\Users\abhagwa\AppData\Local\autonomics\autonomics\Cache/46a42cb13d9_billing19.rnacounts.txt</span>
<span class="co">##      Infer subgroup from sample_ids</span>
<span class="co">##      Infer subgroup from sample_ids</span>
<span class="co">##      Preprocess</span>
<span class="co">##          Keep 22855/56269 features: count &gt;= 10 (~0 + subgroup)</span>
<span class="co">##          pseudocount 0.5</span>
<span class="co">##          cpm:    tmm scale libsizes</span>
<span class="co">##              cpm</span>
<span class="co">##          voom: ~0 + subgroup</span>
<span class="co">##          log2</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + subgroup, weights = assays(object)$weights)</span>
<span class="co">##              contrast ndown  nup</span>
<span class="co">##          1: E00.8-E00     8   36</span>
<span class="co">##          2: E01-E00.8     0    0</span>
<span class="co">##          3:   E02-E01     0   31</span>
<span class="co">##          4:   E05-E02   124   18</span>
<span class="co">##          5:   E15-E05  3179 4165</span>
<span class="co">##          6:   E30-E15  2104 1198</span>
<span class="co">##          7:   M00-E30  1579 1523</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-2-1.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-2-2.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-2-3.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-2-4.png" width="700"></p>
</div>
<div id="read_rnaseq_countssfile" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#read_rnaseq_countssfile" class="anchor"></a>read_rnaseq_counts(sfile)</h3>
<p>The <code>sfile/sfileby</code> arguments in both importers allow to merge in sample data from a separate file by a specified column. This is especially useful in clinical datasets, such as the GSE161731 Covid-19 dataset, for which both a <strong>counts</strong> file as well as a <strong>sample</strong> file can be downloaded from GEO:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">basedir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">'datasets'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">basedir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">basedir</span>, <span class="st">'GSE161731'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu">GEOquery</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GEOquery/man/getGEOSuppFiles.html">getGEOSuppFiles</a></span><span class="op">(</span><span class="st">"GSE161731"</span>, baseDir<span class="op">=</span><span class="va">basedir</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">## Setting options('download.file.method.GEOquery'='auto')</span>
<span class="co">## Setting options('GEOquery.inmemory.gpl'=FALSE)</span>
<span class="co">##                                                                                                          size</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz       8347405</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz      2443</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_key.csv.gz             2398</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_nlcpm.csv.gz   16839511</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_tpm_geo.txt.gz 23859414</span>
<span class="co">##                                                                                                      isdir</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz      FALSE</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz  FALSE</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_key.csv.gz         FALSE</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_nlcpm.csv.gz   FALSE</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_tpm_geo.txt.gz FALSE</span>
<span class="co">##                                                                                                      mode</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz       666</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz   666</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_key.csv.gz          666</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_nlcpm.csv.gz    666</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_tpm_geo.txt.gz  666</span>
<span class="co">##                                                                                                                    mtime</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz      2021-03-23 19:21:39</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz  2021-03-23 19:21:40</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_key.csv.gz         2021-03-23 19:21:40</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_nlcpm.csv.gz   2021-03-23 19:21:48</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_tpm_geo.txt.gz 2021-03-23 19:21:59</span>
<span class="co">##                                                                                                                    ctime</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz      2021-03-23 19:21:33</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz  2021-03-23 19:21:39</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_key.csv.gz         2021-03-23 19:21:40</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_nlcpm.csv.gz   2021-03-23 19:21:40</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_tpm_geo.txt.gz 2021-03-23 19:21:48</span>
<span class="co">##                                                                                                                    atime</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz      2021-03-23 19:21:39</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz  2021-03-23 19:21:40</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_key.csv.gz         2021-03-23 19:21:40</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_nlcpm.csv.gz   2021-03-23 19:21:48</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_tpm_geo.txt.gz 2021-03-23 19:21:59</span>
<span class="co">##                                                                                                      exe</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz       no</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz   no</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_key.csv.gz          no</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_nlcpm.csv.gz    no</span>
<span class="co">## C:\\Users\\abhagwa\\AppData\\Local\\Temp\\RtmpOgH3zs/datasets/GSE161731/GSE161731_xpr_tpm_geo.txt.gz  no</span>
<span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_rnaseq_bams.html">read_rnaseq_counts</a></span><span class="op">(</span>
  file  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">basedir</span>, <span class="st">'GSE161731/GSE161731_counts.csv.gz'</span><span class="op">)</span>, 
  sfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">basedir</span>, <span class="st">'GSE161731/GSE161731_counts_key.csv.gz'</span><span class="op">)</span>, 
  sfileby <span class="op">=</span> <span class="st">'rna_id'</span>, 
  subgroupvar <span class="op">=</span> <span class="st">'gender'</span>, 
  pca <span class="op">=</span> <span class="cn">TRUE</span>, 
  fit <span class="op">=</span> <span class="st">'limma'</span>, 
  plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">##  Read C:\Users\abhagwa\AppData\Local\Temp\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts.csv.gz</span>
<span class="co">##      Merge sdata: C:\Users\abhagwa\AppData\Local\Temp\RtmpOgH3zs/datasets/GSE161731/GSE161731_counts_key.csv.gz</span>
<span class="co">##      Preprocess</span>
<span class="co">##          Keep 18974/60675 features: count &gt;= 10 (~0 + gender)</span>
<span class="co">##          pseudocount 0.5</span>
<span class="co">##          cpm:    tmm scale libsizes</span>
<span class="co">##              cpm</span>
<span class="co">##          voom: ~0 + gender</span>
<span class="co">##          log2</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + gender, weights = assays(object)$weights)</span>
<span class="co">##                contrast ndown nup</span>
<span class="co">##          1: Male-Female    62  61</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-3-1.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-3-2.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-3-3.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-3-4.png" width="700"></p>
</div>
<div id="read_rnaseq_countsblock" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#read_rnaseq_countsblock" class="anchor"></a>read_rnaseq_counts(block)</h3>
<p>The <code>block</code> argument in both importers allows to model e.g. subject_id as a random effect, which is very common in clinical data. Repeating the GEO Covid-19 analysis with subject_id as a block effect. Outcomment the line below, which takes a few minutes to complete.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#object &lt;- read_rnaseq_counts(</span>
<span class="co">#  file  = '~/autonomicscache/datasets/GSE161731/GSE161731_counts.csv.gz', </span>
<span class="co">#  sfile = '~/autonomicscache/datasets/GSE161731/GSE161731_counts_key.csv.gz', </span>
<span class="co">#  sfileby = 'rna_id', </span>
<span class="co">#  subgroupvar = 'gender', </span>
<span class="co">#  block = 'subject_id')</span></code></pre></div>
</div>
<div id="read_rnaseq_bams" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#read_rnaseq_bams" class="anchor"></a>read_rnaseq_bams</h3>
<p><strong>Billing et al. (2016)</strong> compared three types of stem cells: embryonic (<strong>E</strong>), embryonic differentiated into mesenchymal (<strong>EM</strong>), and bone-marrow mesenchymal (<strong>BM</strong>). Importing a downsized version of the RNAseq BAM files (with only a subset of all reads):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">'Rsubread'</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_rnaseq_bams.html">read_rnaseq_bams</a></span><span class="op">(</span>
                dir    <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing16.bam.zip'</span><span class="op">)</span>, 
                paired <span class="op">=</span> <span class="cn">TRUE</span>, 
                genome <span class="op">=</span> <span class="st">'hg38'</span>, 
                pca    <span class="op">=</span> <span class="cn">TRUE</span>, 
                fit   <span class="op">=</span> <span class="st">'limma'</span>, 
                plot   <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">## Lade nötigen Namensraum: Rsubread</span>
<span class="co">## unzip</span>
<span class="co">## Lade nötigen Namensraum: org.Hs.eg.db</span>
<span class="co">## </span>
<span class="co">##      Infer subgroup from sample_ids</span>
<span class="co">##      Preprocess</span>
<span class="co">##          Keep 57/28395 features: count &gt;= 10 (~0 + subgroup)</span>
<span class="co">##          pseudocount 0.5</span>
<span class="co">##          cpm:    tmm scale libsizes</span>
<span class="co">##              cpm</span>
<span class="co">##          voom: ~0 + subgroup</span>
<span class="co">##          log2</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + subgroup, weights = assays(object)$weights)</span>
<span class="co">##             contrast ndown nup</span>
<span class="co">##          1:     E-BM    11  12</span>
<span class="co">##          2:     EM-E    12  10</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-5-1.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-5-2.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-5-3.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-5-4.png" width="700"></p>
</div>
<div id="cpmtmmvoom-effects-on-power" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#cpmtmmvoom-effects-on-power" class="anchor"></a>cpm/tmm/voom effects on power</h3>
<p>Proper preprocessing leads to increased power:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing19.rnacounts.txt'</span><span class="op">)</span>
<span class="co"># log2counts</span>
  <span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_rnaseq_bams.html">read_rnaseq_counts</a></span><span class="op">(</span><span class="va">file</span>, 
       cpm <span class="op">=</span> <span class="cn">FALSE</span>, voom <span class="op">=</span> <span class="cn">FALSE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">##      Infer subgroup from sample_ids</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="../reference/summarize_fit.html">summarize_fit</a></span><span class="op">(</span><span class="va">object</span>, <span class="st">'limma'</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="co">## ndown   nup </span>
<span class="co">##   456    61</span>
<span class="co"># log2cpm</span>
  <span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_rnaseq_bams.html">read_rnaseq_counts</a></span><span class="op">(</span><span class="va">file</span>,
       cpm <span class="op">=</span> <span class="cn">TRUE</span>,  voom <span class="op">=</span> <span class="cn">FALSE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">##      Infer subgroup from sample_ids</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="../reference/summarize_fit.html">summarize_fit</a></span><span class="op">(</span><span class="va">object</span>, <span class="st">'limma'</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="co">## ndown   nup </span>
<span class="co">##  5458  5107</span>
<span class="co"># log2cpm + voom</span>
  <span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_rnaseq_bams.html">read_rnaseq_counts</a></span><span class="op">(</span><span class="va">file</span>,  <span class="co"># log2 cpm + voom</span>
       cpm <span class="op">=</span> <span class="cn">TRUE</span>,  voom <span class="op">=</span> <span class="cn">TRUE</span>,  fit <span class="op">=</span> <span class="st">'limma'</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">##      Infer subgroup from sample_ids</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="../reference/summarize_fit.html">summarize_fit</a></span><span class="op">(</span><span class="va">object</span>, <span class="st">'limma'</span><span class="op">)</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="co">## ndown   nup </span>
<span class="co">##  6994  6971</span></code></pre></div>
</div>
</div>
<div id="proteingroupsphosphosites" class="section level2">
<h2 class="hasAnchor">
<a href="#proteingroupsphosphosites" class="anchor"></a>Proteingroups/phosphosites</h2>
<p>A popular approach in (DDA) MS proteomics data analysis is to use <strong>MaxQuant</strong> to produce <strong>proteinGroups.txt</strong> and <strong>phospho (STY)Sites.txt</strong> files. These can be imported using <code>read_proteingroups</code> / <code>read_phosphosites</code>, which :</p>
<ul>
<li>
<strong>Read</strong>
<ul>
<li>
<strong>quantification cols</strong>: LFQ intensities, Normalized Ratios, or Reporter intensities</li>
<li>
<strong>feature cols</strong>: id, Majority protein IDs, Protein names, Gene names, Contaminant, Potential contaminant, Reverse</li>
<li>
<strong>subgroups</strong> from sampleids / samplefile</li>
</ul>
</li>
<li>
<strong>Preprocess</strong>
<ul>
<li>Convert <strong>0 -&gt; NA</strong> (they represent lack of detection)</li>
<li>
<strong>Log2</strong> transform</li>
<li>Rm <strong>contaminants/reverse/unreplicated</strong> features</li>
<li>
<strong>Impute</strong> consistent NAs with values drawn from a half-normal distribution around zero</li>
<li>For phosphosites: <strong>compute occupancies</strong> by subtracting protein expressions</li>
</ul>
</li>
<li>
<strong>Analyze</strong>
<ul>
<li>plot <strong>sample distributions</strong>
</li>
<li>plot <strong>detections</strong>
</li>
<li>
<strong>pca</strong> biplot</li>
<li>
<strong>Fit</strong> a linear model</li>
<li>Quantify <strong>contrasts</strong> between subsequent factor levels and plot <strong>volcanoes</strong>
</li>
</ul>
</li>
</ul>
<div id="lfq-intensities" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#lfq-intensities" class="anchor"></a>LFQ intensities</h3>
<p>An <strong>LFQ intensities</strong> example is the <strong>Fukuda et al. (2020)</strong> dataset, which compares the proteome of <strong>30d</strong> zebrafish juveniles versus <strong>adults</strong> using label-free quantification. In the volcano plot measured values are shown with circles, imputed values as triangles.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_maxquant.html">read_proteingroups</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'fukuda20.proteingroups.txt'</span><span class="op">)</span>, 
                             pca <span class="op">=</span> <span class="cn">TRUE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">##  Read C:\Users\abhagwa\AppData\Local\autonomics\autonomics\Cache/46a4255b2df6_fukuda20.proteingroups.txt</span>
<span class="co">##      Infer subgroup from sample_ids</span>
<span class="co">##      Replace 0-&gt;NA for 15654/41112 values (in 3998/6852 features of 6/6 samples)</span>
<span class="co">##      Log2 transform</span>
<span class="co">##  Filter features</span>
<span class="co">##      Retain 6735/6852 features: ~Reverse != "+"</span>
<span class="co">##      Retain 6692/6735 features: contaminant != '+'</span>
<span class="co">##      Retain 6679/6692 features: non-zero, non-NA, and non-NaN for some sample</span>
<span class="co">##      Filter 4534/6679 features: expr &gt; 0 for at least two samples in some subgroup</span>
<span class="co">##  Transform exprs</span>
<span class="co">##      Impute systematic nondetects for 762/4534 features in 6/6 samples</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + subgroup)</span>
<span class="co">##                 contrast ndown nup</span>
<span class="co">##          1: Adult-X30dpt  1619 876</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-7-1.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-7-2.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-7-3.png" width="700"></p>
</div>
<div id="normalized-ratios" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#normalized-ratios" class="anchor"></a>Normalized ratios</h3>
<p><strong>Normalized ratios</strong> were used in the proteomics profiling of the earlier described <strong>Billing et al. (2019)</strong> study, which examined the differentiation of embryonic stem cells (<strong>E00</strong>) into mesenchymal stem cells (<strong>E01</strong>,<strong>E02</strong>, <strong>E05</strong>, <strong>E15</strong>, <strong>E30</strong>), and compared these to bone-marrow derived mesenchymal stem cells (<strong>M00</strong>). The proteomics profiling experiment was performed using MS1 labeling: a light (<strong>L</strong>) labeled internal standard was created by mixing all samples in equal amounts, the subsequent samples were either medium (<strong>M</strong>) or heavy (<strong>H</strong>) labeled. Not all label combinations were of interest, and the <code>select_subgroups</code> argument allows to specifies which subgroup combinations to retain:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_maxquant.html">read_proteingroups</a></span><span class="op">(</span>
              file <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing19.proteingroups.txt'</span><span class="op">)</span>,
              select_subgroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'E00_STD'</span>, <span class="st">'E01_STD'</span>, <span class="st">'E02_STD'</span>,
                                   <span class="st">'E05_STD'</span>, <span class="st">'E15_STD'</span>, <span class="st">'E30_STD'</span>, <span class="st">'M00_STD'</span><span class="op">)</span>, 
              pca <span class="op">=</span> <span class="cn">TRUE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">##  Read C:\Users\abhagwa\AppData\Local\autonomics\autonomics\Cache/46a47b883e9_billing19.proteingroups.txt</span>
<span class="co">##      Infer subgroup from sample_ids</span>
<span class="co">##          Retain 21/33samples: ~subgroup %in% select_subgroups</span>
<span class="co">##      Replace NaN-&gt;NA for 77204/205338 values (in 5423/9778 features of 21/21 samples)</span>
<span class="co">##      Log2 transform</span>
<span class="co">##  Filter features</span>
<span class="co">##      Retain 9566/9778 features: ~Reverse != "+"</span>
<span class="co">##      Retain 9470/9566 features: contaminant != '+'</span>
<span class="co">##      Retain 7864/9470 features: non-zero, non-NA, and non-NaN for some sample</span>
<span class="co">##      Filter 7094/7864 features: expr != 0 for at least two samples in some subgroup</span>
<span class="co">##  Transform exprs</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + subgroup)</span>
<span class="co">##                    contrast ndown  nup</span>
<span class="co">##          1: E01_STD-E00_STD     1    2</span>
<span class="co">##          2: E02_STD-E01_STD     1    2</span>
<span class="co">##          3: E05_STD-E02_STD     4    6</span>
<span class="co">##          4: E15_STD-E05_STD  1350 1454</span>
<span class="co">##          5: E30_STD-E15_STD   821  405</span>
<span class="co">##          6: M00_STD-E30_STD  1705  579</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-8-1.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-8-2.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-8-3.png" width="700"></p>
<p>The phosphosites can be read in a similar fashion:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_maxquant.html">read_phosphosites</a></span><span class="op">(</span>
              file <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing19.phosphosites.txt'</span><span class="op">)</span>, 
              proteinfile <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'billing19.proteingroups.txt'</span><span class="op">)</span>, 
              select_subgroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'E00_STD'</span>, <span class="st">'E01_STD'</span>, <span class="st">'E02_STD'</span>,
                                   <span class="st">'E05_STD'</span>, <span class="st">'E15_STD'</span>, <span class="st">'E30_STD'</span>, <span class="st">'M00_STD'</span><span class="op">)</span>, 
              pca <span class="op">=</span> <span class="cn">TRUE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">##  Read C:\Users\abhagwa\AppData\Local\autonomics\autonomics\Cache/46a47b883e9_billing19.proteingroups.txt</span>
<span class="co">##      Infer subgroup from sample_ids</span>
<span class="co">##          Retain 21/33samples: ~subgroup %in% select_subgroups</span>
<span class="co">##      Replace NaN-&gt;NA for 77204/205338 values (in 5423/9778 features of 21/21 samples)</span>
<span class="co">##      Log2 transform</span>
<span class="co">##  Read C:\Users\abhagwa\AppData\Local\autonomics\autonomics\Cache/28e05676396b_billing19.phosphosites.txt</span>
<span class="co">##      Infer subgroup from sample_ids</span>
<span class="co">##          Retain 21/33samples: ~subgroup %in% select_subgroups</span>
<span class="co">##      Replace NaN-&gt;NA for 11144/296226 values (in 1290/14106 features of 21/21 samples)</span>
<span class="co">##      Log2 transform</span>
<span class="co">##  Filter features</span>
<span class="co">##      Retain 13968/14106 features: ~Reverse != "+"</span>
<span class="co">##      Retain 13826/13968 features: contaminant != '+'</span>
<span class="co">##      Retain 9691/13826 features: non-zero, non-NA, and non-NaN for some sample</span>
<span class="co">##      Filter 7072/9691 features: expr != 0 for at least two samples in some subgroup</span>
<span class="co">##      Retain 6034/7072 features: ~`Localization prob` &gt;= min_localization_prob</span>
<span class="co">##  Add occupancies(phospho) = values(phospho) - values(proteins)</span>
<span class="co">##  Transform exprs</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + subgroup)</span>
<span class="co">##                    contrast ndown  nup</span>
<span class="co">##          1: E01_STD-E00_STD     0    4</span>
<span class="co">##          2: E02_STD-E01_STD     1    3</span>
<span class="co">##          3: E05_STD-E02_STD    31   20</span>
<span class="co">##          4: E15_STD-E05_STD  1000 1085</span>
<span class="co">##          5: E30_STD-E15_STD   649  477</span>
<span class="co">##          6: M00_STD-E30_STD  1338  224</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-9-1.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-9-2.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-9-3.png" width="700"></p>
</div>
</div>
<div id="metabolon" class="section level2">
<h2 class="hasAnchor">
<a href="#metabolon" class="anchor"></a>Metabolon</h2>
<p><strong>Metabolon</strong> delivers metabolomics results in the form of an <strong>xlsx</strong> file, with values of interest likely in the (second) <strong>OrigScale</strong> sheet. Features are laid out in rows, samples in columns, and additional feature/sample data are self-contained in the file. Metabolon data can be read/analyzed with the autonomics function <code>read_metabolon</code>, as illustrated below on the dataset of <strong>Halama and and Atkin (2018)</strong>, who investigates the effects of <strong>hypoglycemia</strong> on a number of subjects (<strong>SUB</strong>) at four different time points (<strong>t0</strong>, <strong>t1</strong>, <strong>t2</strong>, <strong>t3</strong>):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_metabolon.html">read_metabolon</a></span><span class="op">(</span>
        file  <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'atkin18.metabolon.xlsx'</span><span class="op">)</span>,  subgroupvar <span class="op">=</span> <span class="st">'SET'</span>,
        pca <span class="op">=</span> <span class="cn">TRUE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, block <span class="op">=</span> <span class="st">'SUB'</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">##  Read: C:\Users\abhagwa\AppData\Local\autonomics\autonomics\Cache/46a459efbda_atkin18.metabolon.xlsx</span>
<span class="co">##      Log2 transform</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + SET | SUB)</span>
<span class="co">##              dupcor `SUB`</span>
<span class="co">## Warning in glmgam.fit(dx, dy, coef.start = start, tol = tol, maxit = maxit, :</span>
<span class="co">## Too much damping - convergence tolerance not achievable</span>
<span class="co">##             contrast ndown nup</span>
<span class="co">##          1:    t1-t0   326  20</span>
<span class="co">##          2:    t2-t1    39  28</span>
<span class="co">##          3:    t3-t2    58 359</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-10-1.png" width="95%"><img src="using_autonomics_files/figure-html/unnamed-chunk-10-2.png" width="95%"><img src="using_autonomics_files/figure-html/unnamed-chunk-10-3.png" width="95%"></p>
</div>
<div id="somascan" class="section level2">
<h2 class="hasAnchor">
<a href="#somascan" class="anchor"></a>Somascan</h2>
<p><strong>Somascan</strong> results are returned in a txt file with extension <strong>.adat</strong>. Features are laid out in columns and samples in rows. Rectangles with feature/sample data are also contained in the file. The autonomics function <code>read_somascan</code> reads/analyzes SOMAscan files, additionally filtering samples/features for quality and type. This is illustrated on the above described dataset from <strong>Halama and and Atkin (2018)</strong>, who investigated the effects of <strong>hypoglycemia</strong> on a number of subjects (<strong>SUB</strong>) at four different time points (<strong>t0</strong>, <strong>t1</strong>, <strong>t2</strong>, <strong>t3</strong>):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/.read_somascan.html">read_somascan</a></span><span class="op">(</span>
            file <span class="op">=</span> <span class="fu"><a href="../reference/download_data.html">download_data</a></span><span class="op">(</span><span class="st">'atkin18.somascan.adat'</span><span class="op">)</span>, 
            pca <span class="op">=</span> <span class="cn">TRUE</span>, fit <span class="op">=</span> <span class="st">'limma'</span>, block <span class="op">=</span> <span class="st">'Subject_ID'</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">##  Read: C:\Users\abhagwa\AppData\Local\autonomics\autonomics\Cache/2d00352b36b8_atkin18.somascan.adat</span>
<span class="co">##      Retain 1125/1129 features: ~ColCheck %in% ~feature_quality</span>
<span class="co">##      Log2 transform</span>
<span class="co">##      Add PCA</span>
<span class="co">##      lmFit(~0 + SampleGroup | Subject_ID)</span>
<span class="co">##              dupcor `Subject_ID`</span>
<span class="co">##             contrast ndown nup</span>
<span class="co">##          1:    t1-t0     1   2</span>
<span class="co">##          2:    t2-t1     2   2</span>
<span class="co">##          3:    t3-t2    65  47</span></code></pre></div>
<p><img src="using_autonomics_files/figure-html/unnamed-chunk-11-1.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-11-2.png" width="700"><img src="using_autonomics_files/figure-html/unnamed-chunk-11-3.png" width="700"></p>
<!--
# complex designs

Multi-factor subgroups are automatically decomposed into `subfactors`, for each of which contrasts between subsequent levels are computed and visualized into a two-dimensional volcano grid.

## The contrastogram
The contrastogram

# Analyzing clinical data
-->
</div>
</div>
<div id="sessioninfo" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>SessionInfo</h1>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## R version 4.0.3 (2020-10-10)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19041)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252   </span>
<span class="co">## [3] LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   </span>
<span class="co">## [5] LC_TIME=German_Germany.1252    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] autonomics_0.99.18 BiocStyle_2.18.1  </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] assertive.base_0.0-7        colorspace_2.0-0           </span>
<span class="co">##   [3] ellipsis_0.3.1              rprojroot_2.0.2            </span>
<span class="co">##   [5] XVector_0.30.0              GenomicRanges_1.42.0       </span>
<span class="co">##   [7] fs_1.5.0                    rstudioapi_0.13            </span>
<span class="co">##   [9] farver_2.0.3                assertive.sets_0.0-3       </span>
<span class="co">##  [11] MultiAssayExperiment_1.16.0 ggrepel_0.9.1              </span>
<span class="co">##  [13] bit64_4.0.5                 AnnotationDbi_1.52.0       </span>
<span class="co">##  [15] xml2_1.3.2                  assertive.data.uk_0.0-2    </span>
<span class="co">##  [17] codetools_0.2-18            R.methodsS3_1.8.1          </span>
<span class="co">##  [19] knitr_1.30                  assertive_0.3-6            </span>
<span class="co">##  [21] assertive.data.us_0.0-2     rematch_1.0.1              </span>
<span class="co">##  [23] dbplyr_2.0.0                R.oo_1.24.0                </span>
<span class="co">##  [25] httr_1.4.2                  BiocManager_1.30.10        </span>
<span class="co">##  [27] readr_1.4.0                 compiler_4.0.3             </span>
<span class="co">##  [29] assertthat_0.2.1            Matrix_1.2-18              </span>
<span class="co">##  [31] limma_3.46.0                htmltools_0.5.1            </span>
<span class="co">##  [33] tools_4.0.3                 gtable_0.3.0               </span>
<span class="co">##  [35] glue_1.4.2                  GenomeInfoDbData_1.2.4     </span>
<span class="co">##  [37] dplyr_1.0.3                 rappdirs_0.3.1             </span>
<span class="co">##  [39] Rcpp_1.0.6                  Biobase_2.50.0             </span>
<span class="co">##  [41] cellranger_1.1.0            pkgdown_1.6.1              </span>
<span class="co">##  [43] vctrs_0.3.6                 assertive.files_0.0-2      </span>
<span class="co">##  [45] assertive.datetimes_0.0-3   assertive.models_0.0-2     </span>
<span class="co">##  [47] xfun_0.20                   stringr_1.4.0              </span>
<span class="co">##  [49] lifecycle_0.2.0             statmod_1.4.35             </span>
<span class="co">##  [51] org.Hs.eg.db_3.12.0         edgeR_3.32.1               </span>
<span class="co">##  [53] zlibbioc_1.36.0             scales_1.1.1               </span>
<span class="co">##  [55] ragg_0.4.1                  pcaMethods_1.82.0          </span>
<span class="co">##  [57] hms_1.0.0                   MatrixGenerics_1.2.0       </span>
<span class="co">##  [59] parallel_4.0.3              SummarizedExperiment_1.20.0</span>
<span class="co">##  [61] GEOquery_2.58.0             assertive.matrices_0.0-2   </span>
<span class="co">##  [63] assertive.strings_0.0-3     yaml_2.2.1                 </span>
<span class="co">##  [65] curl_4.3                    memoise_1.1.0              </span>
<span class="co">##  [67] gridExtra_2.3               ggplot2_3.3.3              </span>
<span class="co">##  [69] stringi_1.5.3               RSQLite_2.2.2              </span>
<span class="co">##  [71] S4Vectors_0.28.1            desc_1.2.0                 </span>
<span class="co">##  [73] BiocGenerics_0.36.0         Rsubread_2.4.2             </span>
<span class="co">##  [75] GenomeInfoDb_1.26.2         rlang_0.4.10               </span>
<span class="co">##  [77] pkgconfig_2.0.3             systemfonts_0.3.2          </span>
<span class="co">##  [79] matrixStats_0.57.0          bitops_1.0-6               </span>
<span class="co">##  [81] evaluate_0.14               lattice_0.20-41            </span>
<span class="co">##  [83] assertive.data_0.0-3        purrr_0.3.4                </span>
<span class="co">##  [85] labeling_0.4.2              assertive.properties_0.0-4 </span>
<span class="co">##  [87] bit_4.0.4                   tidyselect_1.1.0           </span>
<span class="co">##  [89] assertive.code_0.0-3        magrittr_2.0.1             </span>
<span class="co">##  [91] bookdown_0.21               R6_2.5.0                   </span>
<span class="co">##  [93] IRanges_2.24.1              generics_0.1.0             </span>
<span class="co">##  [95] DelayedArray_0.16.0         DBI_1.1.1                  </span>
<span class="co">##  [97] withr_2.4.0                 pillar_1.4.7               </span>
<span class="co">##  [99] assertive.numbers_0.0-2     abind_1.4-5                </span>
<span class="co">## [101] RCurl_1.98-1.2              tibble_3.0.5               </span>
<span class="co">## [103] crayon_1.3.4                assertive.types_0.0-3      </span>
<span class="co">## [105] BiocFileCache_1.14.0        rmarkdown_2.6              </span>
<span class="co">## [107] locfit_1.5-9.4              grid_4.0.3                 </span>
<span class="co">## [109] readxl_1.3.1                data.table_1.13.6          </span>
<span class="co">## [111] blob_1.2.1                  digest_0.6.27              </span>
<span class="co">## [113] tidyr_1.1.2                 R.utils_2.10.1             </span>
<span class="co">## [115] textshaping_0.2.1           stats4_4.0.3               </span>
<span class="co">## [117] munsell_0.5.0               assertive.reflection_0.0-5</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>A M Billing, S S Dib, A M Bhagwat, I T da Silva, R D Drummond, S Hayat, R Al-Mismar, H Ben-Hamidane, N Goswami, K Engholm-Keller, M R Larsen, K Suhre, A Rafii, J Graummann (2019). Mol Cell Proteomics. 18(10):1950-1966. doi: 10.1074/mcp.RA119.001356.</p>
<p>A M Billing, H Ben Hamidane, S S Dib, R J Cotton, A M Bhagwat, P Kumar, S Hayat, N A Yousri, N Goswami, K Suhre, A Rafii, J Graumann (2016). Comprehensive transcriptomics and proteomics characterization of human mesenchymal stem cells reveals source specific cellular markers. Sci Rep. 9;6:21507. doi: 10.1038/srep21507.</p>
<p>R Fukuda, R Marin-Juez, H El-Sammak, A Beisaw, R Ramadass, C Kuenne, S Guenther, A Konzer, A M Bhagwat, J Graumann, D YR Stainier (2020). EMBO Rep. 21(8): e49752. doi: 10.15252/embr.201949752</p>
<p>A Halama, M Kulinski, S Dib, S B Zaghlool, K S Siveen, A Iskandarani, J Zierer 3, K S Prabhu, N J Satheesh, A M Bhagwat, S Uddin, G Kastenmüller, O Elemento, S S Gross, K Suhre (2018). Accelerated lipid catabolism and autophagy are cancer survival mechanisms under inhibited glutaminolysis. Cancer Lett., 430:133-147. <a href="doi:10.1016/j.canlet.2018.05.017" class="uri">doi:10.1016/j.canlet.2018.05.017</a></p>
<p>A Halama, H Kahal, A M Bhagwat, J Zierer, T Sathyapalan, J Graumann, K Suhre, S L Atkin (2018). Metabolic and proteomics signatures of hypoglycaemia in type 2 diabetes. Diabetes, obesity and metabolism, <a href="https://doi.org/10.1111/dom.13602" class="uri">https://doi.org/10.1111/dom.13602</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aditya Bhagwat, Shahina Hayat, Johannes Graumann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
